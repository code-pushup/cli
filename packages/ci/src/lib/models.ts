import type { SourceFileIssue } from './issues';
import type { MonorepoTool } from './monorepo';

/**
 * Customization options for {@link runInCI}
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#options
 */
export type Options = {
  monorepo?: boolean | MonorepoTool;
  projects?: string[] | null;
  task?: string;
  bin?: string;
  config?: string | null;
  directory?: string;
  silent?: boolean;
  debug?: boolean;
  detectNewIssues?: boolean;
  logger?: Logger;
};

/**
 * {@link Options} with filled-in defaults.
 */
export type Settings = Required<Options>;

/**
 * Branches/commits for {@link runInCI}
 */
export type GitRefs = {
  head: GitBranch;
  base?: GitBranch;
};

/**
 * API client for {@link runInCI} - adapter for CI provider
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#provider-api-client
 */
export type ProviderAPIClient = {
  maxCommentChars: number;
  downloadReportArtifact?: () => Promise<string | null>;
  listComments: () => Promise<Comment[]>;
  updateComment: (id: number, body: string) => Promise<Comment>;
  createComment: (body: string) => Promise<Comment>;
};

/**
 * PR comment from {@link ProviderAPIClient}
 */
export type Comment = {
  id: number;
  body: string;
  url: string;
};

/**
 * Git ref (e.g. 'main') and commit sha (e.g. '9d7568265bd5b17d1bb51a6eb6407e0d387058d2')
 */
export type GitBranch = {
  ref: string;
  sha: string;
};

/**
 * Logger instance (e.g. `console`) for reporting progress and problems
 */
export type Logger = {
  error: (message: string) => void;
  warn: (message: string) => void;
  info: (message: string) => void;
  debug: (message: string) => void;
};

/**
 * Resolved return value of {@link runInCI}
 */
export type RunResult = StandaloneRunResult | MonorepoRunResult;

/**
 * Resolved return value of {@link runInCI} in standalone mode
 */
export type StandaloneRunResult = Omit<ProjectRunResult, 'name'> & {
  mode: 'standalone';
  commentId?: number;
};

/**
 * Resolved return value of {@link runInCI} in monorepo mode
 */
export type MonorepoRunResult = {
  mode: 'monorepo';
  projects: ProjectRunResult[];
  commentId?: number;
  diffArtifact?: ArtifactData;
};

/**
 * Result of {@link runInCI} for a single project
 */
export type ProjectRunResult = {
  name: string;
  artifacts: {
    report: ArtifactData;
    diff?: ArtifactData;
  };
  newIssues?: SourceFileIssue[];
};

/**
 * Paths to output files from {@link runInCI}, for uploading as job artifact
 */
export type ArtifactData = {
  rootDir: string;
  files: string[];
};
