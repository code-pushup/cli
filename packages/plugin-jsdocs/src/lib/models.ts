import type { CoverageType } from './runner/models.js';

export type AuditSlug = `${CoverageType}-coverage`;
