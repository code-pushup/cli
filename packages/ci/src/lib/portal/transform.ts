import type {
  AuditFragment,
  BasicTreeNodeDataFragment,
  CategoryFragment,
  CommitFragment,
  CoverageTreeNodeDataFragment,
  GroupFragment,
  IssueFragment,
  PluginFragment,
  ReportFragment,
  TableFragment,
  TreeFragment,
} from '@code-pushup/portal-client';
import type {
  AuditReport,
  BasicTreeNode,
  CategoryConfig,
  CategoryRef,
  Commit,
  CoverageTreeNode,
  Group,
  Issue,
  PluginReport,
  Report,
  Table,
  Tree,
} from '@code-pushup/models';
import { lowercase } from '@code-pushup/utils';

export function transformGQLReport(report: ReportFragment): Report {
  return {
    commit: transformGQLCommit(report.commit),
    plugins: report.plugins.map(plugin =>
      transformGQLPlugin(
        plugin,
        report.issues?.edges.map(({ node }) => node) ?? [],
      ),
    ),
    categories: report.categories.map(transformGQLCategory),
    // TODO: make report metadata required in Portal API?
    packageName: report.packageName ?? 'unknown',
    version: report.packageVersion ?? 'unknown',
    date: report.commandStartDate ?? '',
    duration: report.commandDuration ?? 0,
  };
}

function transformGQLCommit(commit: CommitFragment): Commit {
  return {
    hash: commit.sha,
    message: commit.message,
    // TODO: make commit author and date required in Portal API?
    author: commit.author
      ? `${commit.author.name} <${commit.author.email}>`
      : 'unknown',
    date: commit.date ? new Date(commit.date) : new Date(),
  };
}

function transformGQLCategory(category: CategoryFragment): CategoryConfig {
  return {
    slug: category.slug,
    title: category.title,
    ...(category.description && { description: category.description }),
    refs: category.refs.map(
      ({ target, weight }): CategoryRef => ({
        type: lowercase(target.__typename),
        plugin: target.plugin.slug,
        slug: target.slug,
        weight,
      }),
    ),
    // TODO: Portal API migration - convert isBinary to scoreTarget for backward compatibility
    // Remove this conversion when Portal API supports scoreTarget (#713)
    scoreTarget: category.isBinary ? 1 : undefined,
  };
}

function transformGQLPlugin(
  plugin: PluginFragment,
  issues: IssueFragment[],
): PluginReport {
  return {
    slug: plugin.slug,
    title: plugin.title,
    icon: plugin.icon,
    ...(plugin.description && { description: plugin.description }),
    ...(plugin.docsUrl && { docsUrl: plugin.docsUrl }),
    audits: plugin.audits.edges.map(({ node }) =>
      transformGQLAudit(
        node,
        issues.filter(
          issue =>
            issue.audit.plugin.slug === plugin.slug &&
            issue.audit.slug === node.slug,
        ),
      ),
    ),
    groups: plugin.groups.map(transformGQLGroup),
    ...(plugin.packageName && { packageName: plugin.packageName }),
    ...(plugin.packageVersion && { version: plugin.packageVersion }),
    // TODO: make plugin metadata required in Portal API?
    date: plugin.runnerStartDate ?? '',
    duration: plugin.runnerDuration ?? 0,
  };
}

function transformGQLGroup(group: GroupFragment): Group {
  return {
    slug: group.slug,
    title: group.title,
    ...(group.description && { description: group.description }),
    refs: group.refs.map(({ target, weight }) => ({
      slug: target.slug,
      weight,
    })),
  };
}

function transformGQLAudit(
  audit: AuditFragment,
  issues: IssueFragment[],
): AuditReport {
  return {
    slug: audit.slug,
    title: audit.title,
    ...(audit.description && { description: audit.description }),
    ...(audit.docsUrl && { docsUrl: audit.docsUrl }),
    score: audit.score,
    value: audit.value,
    ...(audit.formattedValue && { displayValue: audit.formattedValue }),
    ...(audit.details?.enabled && {
      details: {
        ...(issues.length > 0 && {
          issues: issues.map(transformGQLIssue),
        }),
        ...(audit.details.tables[0] && {
          table: transformGQLTable(audit.details.tables[0]),
        }),
        ...(audit.details.trees.length > 0 && {
          trees: audit.details.trees.map(transformGQLTree),
        }),
      },
    }),
  };
}

function transformGQLIssue(issue: IssueFragment): Issue {
  return {
    message: issue.message,
    severity: lowercase(issue.severity),
    ...(issue.source?.__typename === 'SourceCodeLocation' && {
      source: {
        file: issue.source.filePath,
        position: {
          startLine: issue.source.startLine ?? 0,
          ...(issue.source.startColumn != null && {
            startColumn: issue.source.startColumn,
          }),
          ...(issue.source.endLine != null && {
            endLine: issue.source.endLine,
          }),
          ...(issue.source.endColumn != null && {
            endColumn: issue.source.endColumn,
          }),
        },
      },
    }),
  };
}

function transformGQLTable(table: TableFragment): Table {
  if (!table.header) {
    return {
      ...(table.title && { title: table.title }),
      rows: table.body.map(cells => cells.map(cell => cell.content)),
    };
  }

  return {
    ...(table.title && { title: table.title }),
    columns: table.header.map(({ content, alignment }, idx) => ({
      key: idx.toString(),
      label: content,
      align: lowercase(alignment),
    })),
    rows: table.body.map(cells =>
      Object.fromEntries(cells.map((cell, idx) => [idx, cell.content])),
    ),
  };
}

function transformGQLTree(tree: TreeFragment): Tree {
  switch (tree.__typename) {
    case 'BasicTree':
      return {
        type: 'basic',
        ...(tree.title && { title: tree.title }),
        root: transformGQLBasicTreeNode(tree.root),
      };
    case 'CoverageTree':
      return {
        type: 'coverage',
        ...(tree.title && { title: tree.title }),
        root: transformGQLCoverageTreeNode(tree.root),
      };
  }
}

// widens limited-depth GraphQL fragment to recursive data structure
type TreeNodeGQL<T> = T & {
  children?: TreeNodeGQL<T>[] | null;
};

function transformGQLBasicTreeNode(
  node: TreeNodeGQL<BasicTreeNodeDataFragment>,
): BasicTreeNode {
  return {
    name: node.name,
    ...(node.customValues && {
      values: Object.fromEntries(
        node.customValues.map(({ key, value }) => [key, value]),
      ),
    }),
    ...(node.children && {
      children: node.children.map(transformGQLBasicTreeNode),
    }),
  };
}

function transformGQLCoverageTreeNode(
  node: TreeNodeGQL<CoverageTreeNodeDataFragment>,
): CoverageTreeNode {
  return {
    name: node.name,
    values: {
      coverage: node.values.coverage,
      ...(node.values.missing && {
        missing: node.values.missing.map(missing => ({
          ...(missing.kind && { kind: missing.kind }),
          ...(missing.name && { name: missing.name }),
          startLine: missing.startLine,
          ...(missing.startColumn && { startColumn: missing.startColumn }),
          ...(missing.endLine && { endLine: missing.endLine }),
          ...(missing.endColumn && { endColumn: missing.endColumn }),
        })),
      }),
    },
    ...(node.children && {
      children: node.children.map(transformGQLCoverageTreeNode),
    }),
  };
}
