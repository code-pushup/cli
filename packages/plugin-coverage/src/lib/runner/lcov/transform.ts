import { LCOVRecord } from 'parse-lcov';
import { AuditOutput, Issue } from '@code-pushup/models';
import { toNumberPrecision, toOrdinal } from '@code-pushup/utils';
import { CoverageType } from '../../config';
import { LCOVStat } from './types';
import { calculateCoverage, mergeConsecutiveNumbers } from './utils';

export function lcovReportToFunctionStat(record: LCOVRecord): LCOVStat {
  return {
    totalFound: record.functions.found,
    totalHit: record.functions.hit,
    issues:
      record.functions.hit < record.functions.found
        ? record.functions.details
            .filter(detail => !detail.hit)
            .map(
              (detail): Issue => ({
                message: `Function ${detail.name} is not called in any test case.`,
                severity: 'error',
                source: {
                  file: record.file,
                  position: { startLine: detail.line },
                },
              }),
            )
        : [],
  };
}

export function lcovReportToLineStat(record: LCOVRecord): LCOVStat {
  const missingCoverage = record.lines.hit < record.lines.found;
  const lines = missingCoverage
    ? record.lines.details
        .filter(detail => !detail.hit)
        .map(detail => detail.line)
    : [];

  const linePositions = mergeConsecutiveNumbers(lines);

  return {
    totalFound: record.lines.found,
    totalHit: record.lines.hit,
    issues: missingCoverage
      ? linePositions.map((linePosition): Issue => {
          const lineReference =
            linePosition.end == null
              ? `Line ${linePosition.start} is`
              : `Lines ${linePosition.start}-${linePosition.end} are`;

          return {
            message: `${lineReference} not covered in any test case.`,
            severity: 'warning',
            source: {
              file: record.file,
              position: {
                startLine: linePosition.start,
                endLine: linePosition.end,
              },
            },
          };
        })
      : [],
  };
}

export function lcovReportToBranchStat(record: LCOVRecord): LCOVStat {
  return {
    totalFound: record.branches.found,
    totalHit: record.branches.hit,
    issues:
      record.branches.hit < record.branches.found
        ? record.branches.details
            .filter(detail => !detail.taken)
            .map(
              (detail): Issue => ({
                message: `${toOrdinal(
                  detail.branch + 1,
                )} branch is not taken in any test case.`,
                severity: 'error',
                source: {
                  file: record.file,
                  position: { startLine: detail.line },
                },
              }),
            )
        : [],
  };
}

export const recordToStatFunctionMapper = {
  branch: lcovReportToBranchStat,
  line: lcovReportToLineStat,
  function: lcovReportToFunctionStat,
};

/**
 *
 * @param stat code coverage result for a given type
 * @param coverageType code coverage type
 * @returns Result of complete code ccoverage data coverted to AuditOutput
 */
export function lcovCoverageToAuditOutput(
  stat: LCOVStat,
  coverageType: CoverageType,
): AuditOutput {
  const coverage = calculateCoverage(stat.totalHit, stat.totalFound);
  const MAX_DECIMAL_PLACES = 4;
  const roundedIntValue = toNumberPrecision(coverage * 100, 0);

  return {
    slug: `${coverageType}-coverage`,
    score: toNumberPrecision(coverage, MAX_DECIMAL_PLACES),
    value: roundedIntValue,
    displayValue: `${roundedIntValue} %`,
    ...(stat.issues.length > 0 && { details: { issues: stat.issues } }),
  };
}
