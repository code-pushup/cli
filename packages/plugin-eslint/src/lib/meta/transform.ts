import type { Audit } from '@code-pushup/models';
import { truncateDescription, truncateTitle } from '@code-pushup/utils';
import { ruleIdToSlug } from './hash';
import { RuleData } from './rules';

export function ruleToAudit({ ruleId, meta, options }: RuleData): Audit {
  const name = ruleId.split('/').at(-1) ?? ruleId;
  const plugin =
    name === ruleId ? null : ruleId.slice(0, ruleId.lastIndexOf('/'));
  const pluginContext = plugin ? `, from _${plugin}_ plugin` : '';

  const lines: string[] = [
    `ESLint rule **${name}**${pluginContext}.`,
    ...(options?.length ? ['Custom options:'] : []),
    ...(options?.map(option =>
      ['```json', JSON.stringify(option, null, 2), '```'].join('\n'),
    ) ?? []),
  ];

  return {
    slug: ruleIdToSlug(ruleId, options),
    title: truncateTitle(meta.docs?.description ?? name),
    description: truncateDescription(lines.join('\n\n')),
    ...(meta.docs?.url && {
      docsUrl: meta.docs.url,
    }),
  };
}
