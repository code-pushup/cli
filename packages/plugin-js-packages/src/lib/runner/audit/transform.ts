import type { AuditOutput, Issue, IssueSeverity } from '@code-pushup/models';
import { objectToEntries } from '@code-pushup/utils';
import {
  DependencyGroup,
  PackageAuditLevel,
  PackageManager,
  packageAuditLevels,
} from '../../config';
import { auditScoreModifiers } from './constants';
import { AuditResult } from './types';

export function auditResultToAuditOutput(
  result: AuditResult,
  packageManager: PackageManager,
  dependenciesType: DependencyGroup,
  auditLevelMapping: Record<PackageAuditLevel, IssueSeverity>,
): AuditOutput {
  const issues = vulnerabilitiesToIssues(
    result.vulnerabilities,
    auditLevelMapping,
  );

  return {
    slug: `${packageManager}-audit-${dependenciesType}`,
    score: calculateAuditScore(result.summary),
    value: result.summary.total,
    displayValue: vulnerabilitiesToDisplayValue(result.summary),
    ...(issues.length > 0 && { details: { issues } }),
  };
}

export function calculateAuditScore(
  stats: Record<PackageAuditLevel | 'total', number>,
) {
  if (stats.total === 0) {
    return 1;
  }

  return objectToEntries(stats).reduce<number>(
    (score, [level, vulnerabilities]) => {
      if (level === 'total') {
        return score;
      }

      const reducedScore = score - auditScoreModifiers[level] * vulnerabilities;
      return Math.max(reducedScore, 0);
    },
    1,
  );
}

export function vulnerabilitiesToDisplayValue(
  vulnerabilities: Record<PackageAuditLevel | 'total', number>,
): string {
  if (vulnerabilities.total === 0) {
    return '0 vulnerabilities';
  }

  const vulnerabilityStats = packageAuditLevels
    .map(level =>
      vulnerabilities[level] > 0 ? `${vulnerabilities[level]} ${level}` : '',
    )
    .filter(text => text !== '')
    .join(', ');
  return `${vulnerabilities.total} ${
    vulnerabilities.total === 1 ? 'vulnerability' : 'vulnerabilities'
  } (${vulnerabilityStats})`;
}

export function vulnerabilitiesToIssues(
  vulnerabilities: AuditResult['vulnerabilities'],
  auditLevelMapping: Record<PackageAuditLevel, IssueSeverity>,
): Issue[] {
  if (vulnerabilities.length === 0) {
    return [];
  }

  return Object.values(vulnerabilities).map<Issue>(detail => {
    const versionRange =
      detail.versionRange === '*'
        ? '**all** versions'
        : `versions **${detail.versionRange}**`;
    const depHierarchy =
      typeof detail.directDependency === 'string'
        ? `\`${detail.directDependency}\`'${
            detail.directDependency.endsWith('s') ? '' : 's'
          } dependency \`${detail.name}\``
        : `\`${detail.name}\` dependency`;

    const vulnerabilitySummary = `has a **${detail.severity}** vulnerability in ${versionRange}.`;
    const fixInfo = detail.fixInformation ? ` ${detail.fixInformation}` : '';
    const additionalInfo =
      detail.title != null && detail.url != null
        ? ` More information: [${detail.title}](${detail.url})`
        : '';

    return {
      message: `${depHierarchy} ${vulnerabilitySummary}${fixInfo}${additionalInfo}`,
      severity: auditLevelMapping[detail.severity],
    };
  });
}
