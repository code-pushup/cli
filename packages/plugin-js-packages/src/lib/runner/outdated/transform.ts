import { ReleaseType, clean, diff, neq } from 'semver';
import type { AuditOutput, Issue } from '@code-pushup/models';
import { objectFromEntries, pluralize } from '@code-pushup/utils';
import { DependencyGroup, PackageManagerId } from '../../config';
import { dependencyGroupToLong } from '../../constants';
import { RELEASE_TYPES, outdatedSeverity } from './constants';
import { OutdatedResult } from './types';

export function outdatedResultToAuditOutput(
  result: OutdatedResult,
  packageManager: PackageManagerId,
  depGroup: DependencyGroup,
  totalDeps: number,
): AuditOutput {
  const relevantDependencies: OutdatedResult = result.filter(
    dep => dep.type === dependencyGroupToLong[depGroup],
  );

  const validDependencies = relevantDependencies
    .map(dep => ({
      ...dep,
      current: clean(dep.current),
      latest: clean(dep.latest),
    }))
    .filter(
      (dep): dep is OutdatedResult[number] =>
        dep.current != null && dep.latest != null,
    );

  const outdatedDependencies = validDependencies.filter(dep =>
    neq(dep.current, dep.latest),
  );

  const outdatedStats = outdatedDependencies.reduce((acc, dep) => {
    const outdatedLevel = diff(dep.current, dep.latest);
    if (outdatedLevel == null) {
      return acc;
    }
    return { ...acc, [outdatedLevel]: acc[outdatedLevel] + 1 };
  }, objectFromEntries(RELEASE_TYPES.map(versionType => [versionType, 0])));

  const issues =
    outdatedDependencies.length === 0
      ? []
      : outdatedToIssues(outdatedDependencies);

  return {
    slug: `${packageManager}-outdated-${depGroup}`,
    score: calculateOutdatedScore(outdatedStats.major, totalDeps),
    value: outdatedDependencies.length,
    displayValue: outdatedToDisplayValue(outdatedStats),
    details: { issues },
  };
}

export function calculateOutdatedScore(
  majorOutdated: number,
  totalDeps: number,
) {
  return totalDeps > 0 ? (totalDeps - majorOutdated) / totalDeps : 1;
}

export function outdatedToDisplayValue(stats: Record<ReleaseType, number>) {
  const total = Object.values(stats).reduce((acc, value) => acc + value, 0);

  const versionBreakdown = RELEASE_TYPES.map(version =>
    stats[version] > 0 ? `${stats[version]} ${version}` : '',
  ).filter(text => text !== '');

  if (versionBreakdown.length === 0) {
    return 'all dependencies are up to date';
  }

  if (versionBreakdown.length > 1) {
    return `${total} outdated package versions (${versionBreakdown.join(
      ', ',
    )})`;
  }

  return `${versionBreakdown[0]} outdated package ${pluralize(
    'version',
    total,
  )}`;
}

export function outdatedToIssues(dependencies: OutdatedResult): Issue[] {
  return dependencies.map<Issue>(dep => {
    const { name, current, latest, url } = dep;
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const outdatedLevel = diff(current, latest)!;
    const packageReference =
      url == null ? `\`${name}\`` : `[\`${name}\`](${url})`;

    return {
      message: `Package ${packageReference} requires a **${outdatedLevel}** update from **${current}** to **${latest}**.`,
      severity: outdatedSeverity[outdatedLevel],
    };
  });
}
