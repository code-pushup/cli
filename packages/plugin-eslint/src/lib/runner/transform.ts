import type { Linter } from 'eslint';
import type { AuditOutput, Issue, IssueSeverity } from '@code-pushup/models';
import {
  compareIssueSeverity,
  countOccurrences,
  objectToEntries,
  pluralizeToken,
  truncateIssueMessage,
} from '@code-pushup/utils';
import { ruleIdToSlug } from '../meta/hash';
import type { LinterOutput } from './types';

type LintIssue = Linter.LintMessage & {
  filePath: string;
};

export function lintResultsToAudits({
  results,
  ruleOptionsPerFile,
}: LinterOutput): AuditOutput[] {
  const issuesPerAudit = results
    .flatMap(({ messages, filePath }) =>
      messages.map((message): LintIssue => ({ ...message, filePath })),
    )
    .reduce<Record<string, LintIssue[]>>((acc, issue) => {
      const { ruleId, message, filePath } = issue;
      if (!ruleId) {
        console.warn(`ESLint core error - ${message}`);
        return acc;
      }
      const options = ruleOptionsPerFile[filePath]?.[ruleId] ?? [];
      const auditSlug = ruleIdToSlug(ruleId, options);
      return { ...acc, [auditSlug]: [...(acc[auditSlug] ?? []), issue] };
    }, {});

  return Object.entries(issuesPerAudit).map(entry => toAudit(...entry));
}

function toAudit(slug: string, issues: LintIssue[]): AuditOutput {
  const auditIssues = issues.map(convertIssue);
  const severityCounts = countOccurrences(
    auditIssues.map(({ severity }) => severity),
  );
  const severities = objectToEntries(severityCounts);
  const summaryText = [...severities]
    .sort((a, b) => -compareIssueSeverity(a[0], b[0]))
    .map(([severity, count = 0]) => pluralizeToken(severity, count))
    .join(', ');

  return {
    slug,
    score: Number(auditIssues.length === 0),
    value: auditIssues.length,
    displayValue: summaryText,
    details: {
      issues: auditIssues,
    },
  };
}

function convertIssue(issue: LintIssue): Issue {
  return {
    message: truncateIssueMessage(issue.message),
    severity: convertSeverity(issue.severity),
    source: {
      file: issue.filePath,
      position: {
        startLine: issue.line,
        startColumn: issue.column,
        endLine: issue.endLine,
        endColumn: issue.endColumn,
      },
    },
  };
}

function convertSeverity(severity: Linter.Severity): IssueSeverity {
  switch (severity) {
    case 2:
      return 'error';
    case 1:
      return 'warning';
    case 0:
      // shouldn't happen
      throw new Error(`Unexpected severity ${severity} in ESLint results`);
  }
}
