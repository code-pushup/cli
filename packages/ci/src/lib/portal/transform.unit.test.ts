import {
  IssueSeverity,
  type ReportFragment,
  TableAlignment,
} from '@code-pushup/portal-client';
import { reportSchema } from '@code-pushup/models';
import { transformGQLReport } from './transform.js';

describe('transformGQLReport', () => {
  const GQL_REPORT: ReportFragment = {
    commit: {
      sha: '4da737d63efcc83d0dd05620801f195968611eb7',
      message: 'Apply suggestions from code review',
      author: {
        name: 'John Doe',
        email: 'john.doe@example.com',
      },
      date: '2025-08-01T00:00:00.000Z',
    },
    packageName: '@code-pushup/core',
    packageVersion: '0.42.0',
    commandStartDate: '2025-08-01T00:10:00.000Z',
    commandDuration: 30_000,
    categories: [
      {
        slug: 'code-style',
        title: 'Code style',
        score: 0.5,
        refs: [
          {
            target: {
              __typename: 'Group',
              plugin: { slug: 'eslint' },
              slug: 'suggestions',
            },
            weight: 1,
          },
        ],
      },
      {
        slug: 'bundle-size',
        title: 'Bundle size',
        score: 0.75,
        refs: [
          {
            target: {
              __typename: 'Audit',
              plugin: { slug: 'bundle-stats' },
              slug: 'initial',
            },
            weight: 1,
          },
        ],
      },
    ],
    plugins: [
      {
        slug: 'eslint',
        title: 'ESLint',
        icon: 'eslint',
        packageName: '@code-pushup/eslint-plugin',
        packageVersion: '0.42.0',
        runnerStartDate: '2025-08-01T00:10:00.000Z',
        runnerDuration: 20_000,
        audits: {
          edges: [
            {
              node: {
                slug: 'max-lines',
                title: 'Enforce a maximum number of lines per file',
                score: 0,
                value: 1,
                formattedValue: '1 warning',
                details: { enabled: true, trees: [], tables: [] },
              },
            },
            {
              node: {
                slug: 'max-lines-per-function',
                title: 'Enforce a maximum number of lines of code per function',
                score: 1,
                value: 0,
                formattedValue: 'passed',
                details: { enabled: true, trees: [], tables: [] },
              },
            },
          ],
        },
        groups: [
          {
            slug: 'suggestions',
            title: 'Suggestion',
            score: 0.5,
            refs: [
              { target: { slug: 'max-lines' }, weight: 1 },
              { target: { slug: 'max-lines-per-function' }, weight: 1 },
            ],
          },
        ],
      },
      {
        slug: 'bundle-stats',
        title: 'Bundle stats',
        icon: 'javascript-map',
        runnerStartDate: '2025-08-01T00:10:20.000Z',
        runnerDuration: 10_000,
        audits: {
          edges: [
            {
              node: {
                slug: 'initial',
                title: 'Initial JavaScript bundle',
                score: 0.75,
                value: 420_000,
                formattedValue: '420 kB',
                details: {
                  enabled: true,
                  tables: [
                    {
                      header: [
                        { content: 'Group', alignment: TableAlignment.Left },
                        { content: 'Size', alignment: TableAlignment.Right },
                        { content: 'Modules', alignment: TableAlignment.Right },
                      ],
                      body: [
                        [
                          {
                            content: '3rd-party packages',
                            alignment: TableAlignment.Left,
                          },
                          {
                            content: '321.4 kB',
                            alignment: TableAlignment.Right,
                          },
                          {
                            content: '101',
                            alignment: TableAlignment.Right,
                          },
                        ],
                        [
                          {
                            content: 'Application shell',
                            alignment: TableAlignment.Left,
                          },
                          {
                            content: '98.6 kB',
                            alignment: TableAlignment.Right,
                          },
                          {
                            content: '7',
                            alignment: TableAlignment.Right,
                          },
                        ],
                      ],
                    },
                  ],
                  trees: [
                    {
                      __typename: 'BasicTree',
                      root: {
                        name: 'stats.json',
                        children: [
                          {
                            name: 'outputs',
                            children: [
                              {
                                name: 'dist/main.js',
                                customValues: [
                                  { key: 'size', value: '420 kB' },
                                ],
                                children: [
                                  {
                                    name: 'inputs',
                                    children: [
                                      {
                                        name: 'src/main.ts',
                                        customValues: [
                                          { key: 'size', value: '275 kB' },
                                        ],
                                      },
                                      {
                                        name: 'src/utils/format.ts',
                                        customValues: [
                                          { key: 'size', value: '120 kB' },
                                        ],
                                      },
                                      {
                                        name: 'src/utils/math.ts',
                                        customValues: [
                                          { key: 'size', value: '25 kB' },
                                        ],
                                      },
                                    ],
                                  },
                                ],
                              },
                            ],
                          },
                        ],
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
        groups: [],
      },
    ],
    issues: {
      edges: [
        {
          node: {
            audit: { plugin: { slug: 'eslint' }, slug: 'max-lines' },
            message: 'File has too many lines (420). Maximum allowed is 300.',
            severity: IssueSeverity.Warning,
            source: {
              __typename: 'SourceCodeLocation',
              filePath: 'src/main.ts',
              startLine: 301,
              endLine: 420,
            },
          },
        },
        {
          node: {
            audit: { plugin: { slug: 'bundle-stats' }, slug: 'initial' },
            message:
              '`main.js` is **420 kB**, exceeds warning threshold of 350 kB',
            severity: IssueSeverity.Warning,
          },
        },
      ],
    },
  };

  it('should convert full GraphQL report to valid report.json format', () => {
    const report = transformGQLReport(GQL_REPORT);
    expect(() => reportSchema.parse(report)).not.toThrow();
    expect(report).toMatchSnapshot();
  });
});
