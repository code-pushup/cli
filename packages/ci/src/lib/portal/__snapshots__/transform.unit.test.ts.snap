// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformGQLReport > should convert full GraphQL report to valid report.json format 1`] = `
{
  "categories": [
    {
      "refs": [
        {
          "plugin": "eslint",
          "slug": "suggestions",
          "type": "group",
          "weight": 1,
        },
      ],
      "slug": "code-style",
      "title": "Code style",
    },
    {
      "refs": [
        {
          "plugin": "bundle-stats",
          "slug": "initial",
          "type": "audit",
          "weight": 1,
        },
      ],
      "slug": "bundle-size",
      "title": "Bundle size",
    },
  ],
  "commit": {
    "author": "John Doe <john.doe@example.com>",
    "date": 2025-08-01T00:00:00.000Z,
    "hash": "4da737d63efcc83d0dd05620801f195968611eb7",
    "message": "Apply suggestions from code review",
  },
  "date": "2025-08-01T00:10:00.000Z",
  "duration": 30000,
  "packageName": "@code-pushup/core",
  "plugins": [
    {
      "audits": [
        {
          "details": {
            "issues": [
              {
                "message": "File has too many lines (420). Maximum allowed is 300.",
                "severity": "warning",
                "source": {
                  "file": "src/main.ts",
                  "position": {
                    "endLine": 420,
                    "startLine": 301,
                  },
                },
              },
            ],
          },
          "displayValue": "1 warning",
          "score": 0,
          "slug": "max-lines",
          "title": "Enforce a maximum number of lines per file",
          "value": 1,
        },
        {
          "details": {},
          "displayValue": "passed",
          "score": 1,
          "slug": "max-lines-per-function",
          "title": "Enforce a maximum number of lines of code per function",
          "value": 0,
        },
      ],
      "date": "2025-08-01T00:10:00.000Z",
      "duration": 20000,
      "groups": [
        {
          "refs": [
            {
              "slug": "max-lines",
              "weight": 1,
            },
            {
              "slug": "max-lines-per-function",
              "weight": 1,
            },
          ],
          "slug": "suggestions",
          "title": "Suggestion",
        },
      ],
      "icon": "eslint",
      "packageName": "@code-pushup/eslint-plugin",
      "slug": "eslint",
      "title": "ESLint",
      "version": "0.42.0",
    },
    {
      "audits": [
        {
          "details": {
            "issues": [
              {
                "message": "\`main.js\` is **420 kB**, exceeds warning threshold of 350 kB",
                "severity": "warning",
              },
            ],
            "table": {
              "columns": [
                {
                  "align": "left",
                  "key": "0",
                  "label": "Group",
                },
                {
                  "align": "right",
                  "key": "1",
                  "label": "Size",
                },
                {
                  "align": "right",
                  "key": "2",
                  "label": "Modules",
                },
              ],
              "rows": [
                {
                  "0": "3rd-party packages",
                  "1": "321.4 kB",
                  "2": "101",
                },
                {
                  "0": "Application shell",
                  "1": "98.6 kB",
                  "2": "7",
                },
              ],
            },
            "trees": [
              {
                "root": {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "name": "src/main.ts",
                                  "values": {
                                    "size": "275 kB",
                                  },
                                },
                                {
                                  "name": "src/utils/format.ts",
                                  "values": {
                                    "size": "120 kB",
                                  },
                                },
                                {
                                  "name": "src/utils/math.ts",
                                  "values": {
                                    "size": "25 kB",
                                  },
                                },
                              ],
                              "name": "inputs",
                            },
                          ],
                          "name": "dist/main.js",
                          "values": {
                            "size": "420 kB",
                          },
                        },
                      ],
                      "name": "outputs",
                    },
                  ],
                  "name": "stats.json",
                },
                "type": "basic",
              },
            ],
          },
          "displayValue": "420 kB",
          "score": 0.75,
          "slug": "initial",
          "title": "Initial JavaScript bundle",
          "value": 420000,
        },
      ],
      "date": "2025-08-01T00:10:20.000Z",
      "duration": 10000,
      "groups": [],
      "icon": "javascript-map",
      "slug": "bundle-stats",
      "title": "Bundle stats",
    },
  ],
  "version": "0.42.0",
}
`;
